### PR TITLE
Feature/add final form redux

### PR DESCRIPTION
**Pull request by marksy**
- updated comment ref for multichoice component
- added description for multichoice component
- removed commented-out code
- fixed conflict in snapshot
- added FinalFormWrapper component added to storybook, example shown in radio

A side note: error in console (_Invalid prop `options` of type `array` supplied to `RadioGroup`_)

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -61,9 +61,11 @@
     "govuk-colours": "^1.0.3"
   },
   "peerDependencies": {
+    "final-form": "^4.2.0",
```
Comment (penx) [inline package.json]: these should be devDependencies not peerDependencies

File: src/components/Label/index.js [hunk 2]
```diff
@@ -11,6 +11,11 @@ const Label = glamorous.label(
     flexDirection: "column",
     boxSizing: "border-box",
     width: "100%",
+    ":after": {
+      content: "''",
```
Comment (penx) [inline src/components/Label/index.js]: it's not clear to me what this css is needed for, please add a comment inline explaining what it's doing

File: src/components/MultiChoice/index.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+const FieldSet = glamorous.div(
+  {
+    padding: 0,
+    margin: 0,
+    border: 0,
+    boxSizing: "border-box",
+    width: "100%",
+    ":after": {
+      content: "''",
```
Comment (penx) [inline src/components/MultiChoice/index.js]: Same here, why `"''"`, did you mean `""`?

Comment (marksy) [reply src/components/MultiChoice/index.js]: in CSS it is: `content: "";` (to have empty content). In CSS-in-JS you need to have single quotes inside the `content` property

File: src/components/Radio/stories.js [hunk 4]
```diff
@@ -1,46 +1,178 @@
...
+
+import Button from "../Button/index";
+import GridCol from "../GridCol/index";
+import GridRow from "../GridRow/index";
+import Layout from "../Layout/index";
+import MultiChoice from "../MultiChoice/index";
 import Radio from ".";
 
+const onSubmit = values => {
+  console.log(JSON.stringify(values, 0, 2));
```
Comment (penx) [inline src/components/Radio/stories.js]: boop

File: src/components/Label/index.js (hunk 2 above)
Comment (marksy) [reply src/components/Label/index.js]: it's a clearfix to allow the label to clear the floated children

File: src/components/Radio/stories.js (hunk 4 above)
Comment (marksy) [reply src/components/Radio/stories.js]: 🤫

File: src/components/MultiChoice/index.js (hunk 3 above)
Comment (penx) [reply src/components/MultiChoice/index.js]: oh right ha

File: src/components/Radio/stories.js [hunk 5]
```diff
@@ -1,46 +1,178 @@
...
+const RadioGroup = ({ label, hint, name, options, inline, ...input }) => (
+  <Layout>
+    <GridRow>
+      <GridCol>
+        <MultiChoice {...input} label={label} hint={hint}>
+          {options.map(o => (
+            <div key={o.value}>
+              <Radio
+                type="radio"
+                {...input}
```
Comment (penx) [inline src/components/Radio/stories.js]: hmm feels weird to pass ...input to both the radio and the multichoice - why is this needed?

File: src/components/Radio/stories.js (hunk 4 above)
Comment (penx) [reply src/components/Radio/stories.js]: you can use "@storybook/addon-actions" for this

File: src/components/Label/index.js (hunk 2 above)
Comment (penx) [reply src/components/Label/index.js]: durp ok no need for comment

Comment (penx) [reply src/components/Label/index.js]: the "''" threw me

File: src/components/Radio/index.js [hunk 6]
```diff
@@ -87,9 +87,23 @@ const LabelText = glamorous.span({
   }
 });
 
-const Radio = ({ children, inline, input }) => (
+const Radio = ({
```
Comment (penx) [inline src/components/Radio/index.js]: I would categorise form components in to one of 2 types:

1. `Field` if it receives both input and meta data from final form, i.e. displays validation message
2. `Input` if it just receives input props

This seems to be an Input.

I would suggest all inputs should receive props directly rather than an `input` object prop.

i.e. parent field/group would pass `<Input {...input} />`

File: src/components/Radio/stories.js [hunk 7]
```diff
@@ -1,46 +1,178 @@
...
+  name: PropTypes.string,
+  inline: PropTypes.bool,
+  options: PropTypes.shape({
+    title: PropTypes.string,
+    value: PropTypes.string
+  })
+};
+
+const App = () => (
+  <div>
```
Comment (penx) [inline src/components/Radio/stories.js]:
```<div>
 +    <Form
 +      onSubmit={onSubmit}
 +      render={({ handleSubmit, reset, submitting, pristine, values }) => (
 +        <form onSubmit={handleSubmit}>
 +          <Layout>
 +            <GridRow>
 +              <GridCol>```

I would suggest creating a util component that does this bit so we can repeat it elsewhere, maybe stick it in /stories/components/FinalFormWraper.js.

File: src/components/Radio/stories.js [hunk 8]
```diff
@@ -1,46 +1,178 @@
...
+    <Form
+      onSubmit={onSubmit}
+      render={({ handleSubmit, reset, submitting, pristine, values }) => (
+        <form onSubmit={handleSubmit}>
+          <Layout>
+            <GridRow>
+              <GridCol>
```
Comment (penx) [inline src/components/Radio/stories.js]: I'd avoid the grid stuff in the final form stories, just use the minimal code needed to show how to hook up to final form

File: src/components/Radio/stories.js [hunk 9]
```diff
@@ -1,46 +1,178 @@
...
+                <Button onClick={reset} disabled={submitting || pristine}>
+                  Reset
+                </Button>
+              </GridCol>
+            </GridRow>
+            <GridRow>
+              <GridCol>
+                <hr />
+                <pre>{JSON.stringify(values, 0, 2)}</pre>
```
Comment (penx) [inline src/components/Radio/stories.js]: may be able to do something nice with `@storybook/addon-actions` instead of showing inline

File: package.json [hunk 10]
```diff
@@ -1,4 +1,4 @@
-{
+ {
```
Comment (penx) [inline package.json]: remove space plox

File: src/components/MultiChoice/index.js [hunk 11]
```diff
@@ -0,0 +1,78 @@
+// TODO INSERT A COMMENT REFERENCE TO EXTERNAL URL IF POSSIBLE
```
Comment (penx) [inline src/components/MultiChoice/index.js]: update this plox

File: src/components/Radio/stories.js [hunk 12]
```diff
@@ -1,7 +1,95 @@
 import React from "react";
+import { Form, Field } from "react-final-form";
 import { storiesOf } from "@storybook/react";
+import { action } from "@storybook/addon-actions";
+import PropTypes from "prop-types";
+
+import Button from "../Button/index";
+// import GridCol from "../GridCol/index";
```
Comment (penx) [inline src/components/Radio/stories.js]: remove unused code

File: src/components/DateInput/__snapshots__/test.js.snap [hunk 13]
```diff
@@ -211,3 +218,226 @@ exports[`DateInput matches snapshot: enzyme.mount 1`] = `
   </glamorous(div)>
 </DateInput>
 `;
+<<<<<<< HEAD
```
Comment (penx) [inline src/components/DateInput/__snapshots__/test.js.snap]: you've committed a conflict

File: src/components/MultiChoice/index.js [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+
+import React from "react";
```
Comment (penx) [inline src/components/MultiChoice/index.js]: also I think this file needs a comment describing what it's expected usage is

File: src/components/Radio/stories.js [hunk 15]
```diff
@@ -31,16 +116,21 @@ storiesOf("Radio", module).add("Radio disabled", () => (
 
 storiesOf("Radio", module).add("Radio preselected", () => (
   <div>
-    <Radio name="group1" defaultChecked>
+    <Radio name="group1" readOnly checked>
```
Comment (penx) [inline src/components/Radio/stories.js]: why readOnly? Would checked  work on its own? Maybe create a separate story for readOnly?

Comment (marksy) [reply src/components/Radio/stories.js]: yeah... this was one of the errors/warnings it displayed said it should be either readOnly when i was trying to get the reset working, so this was just trying to please that error

File: src/stories/components/FinalFormWrapper.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+
+const FinalFormWrapper = ({ children }) => (
+  <Form
+    onSubmit={action("submit")}
+    render={({ handleSubmit, reset, submitting, pristine, values }) => (
+      <form onSubmit={handleSubmit}>
+        <div>{children}</div>
+        <div>
+          <Button type="submit" disabled={submitting}>
+            Log In
```
Comment (penx) [inline src/stories/components/FinalFormWrapper.js]: Change this to "Submit" init

File: .eslintrc [hunk 17]
```diff
@@ -11,11 +11,11 @@
   "rules": {
     "prettier/prettier": "error",
     "react/jsx-filename-extension": 0,
-    "jsx-a11y/label-has-for": 0,
+    "jsx-a11y/label-has-for": 0
   },
   "overrides": [
     {
-      "files": [ "stories.js", "test.js", "src/stories/index.js", "src/setupTests.js", "scripts/createComponent.js" ],
+      "files": [ "stories.js", "test.js", "src/stories/index.js", "src/setupTests.js", "scripts/createComponent.js", "src/stories/components/FinalFormWrapper.js" ],
```
Comment (penx) [inline .eslintrc]: just have "src/stories" init

File: src/components/Radio/stories.js [hunk 18]
```diff
@@ -1,7 +1,55 @@
 import React from "react";
+import { Field } from "react-final-form";
 import { storiesOf } from "@storybook/react";
+// import { action } from "@storybook/addon-actions";
```
Comment (penx) [inline src/components/Radio/stories.js]: remove commented out code
